src/agent.jl

```diff
@@ -37,60 +37,59 @@ function iterate_agents_cb(agent::HSA.Agent, agents)
     return HSA.STATUS_SUCCESS
 end
 
-"Determines if a memory region can be used for kernarg allocations."
-function get_kernarg_memory_region_cb(region::HSA.Region, data::Ptr{HSA.Region})
+@inline function get_memory_region(region::HSA.Region, data::Ptr{HSA.Region}, kind, prefer_device=true)
     segment = Ref{HSA.RegionSegment}()
     HSA.region_get_info(region, HSA.REGION_INFO_SEGMENT, segment)
+    # Reject non-global regions
     if (segment[] != HSA.REGION_SEGMENT_GLOBAL)
         return HSA.STATUS_SUCCESS
     end
 
-    flags = Ref{HSA.RegionGlobalFlag}()
-    HSA.region_get_info(region, HSA.REGION_INFO_GLOBAL_FLAGS, flags)
-    if (flags[] & HSA.REGION_GLOBAL_FLAG_KERNARG > 0)
+    r_flags = Ref{HSA.RegionGlobalFlag}()
+    HSA.region_get_info(region, HSA.REGION_INFO_GLOBAL_FLAGS, r_flags)
+    flags = r_flags[]
+    
+    r_host_accessible_region = Ref(false)
+    HSA.region_get_info(region,
+        reinterpret(HSA.LibHSARuntime.hsa_region_info_t, HSA.AMD_REGION_INFO_HOST_ACCESSIBLE),
+        r_host_accessible_region)
+    host_accessible = r_host_accessible_region[]
+    if flags & kind > 0
+        # If we are looking for a coarse region and we are prefering device local, skip the host accessible region
+        # following https://cs.github.com/ROCm-Developer-Tools/LLVM-AMDGPU-Assembler-Extra/blob/844212f486868fe5c43cdf8a4151d11a879c74ba/examples/common/dispatch.cpp#L112
+        if kind == HSA.REGION_GLOBAL_FLAG_COARSE_GRAINED
+            if host_accessible == prefer_device
```

Comment:
I feel like it would make more sense to have the kwarg be `allow_host_access`, and then:
```suggestion
            if !allow_host_access && host_accessible
```

Reply:
☝️ +1